nixos/modules/services/web-apps/keycloak.nix

```
@@ -503,57 +555,65 @@ in
                     ""
                   else
                     throw "Unsupported type '${type}' for attribute '${attribute}'!";
+
             in
-              lib.concatStringsSep ", " (lib.mapAttrsToList makeArg set);
+              concatStringsSep ", " (mapAttrsToList makeArg set);
 
 
-          /* Recurses into the `attrs` attrset, beginning at the path
-             resolved from `state.path ++ node`; if `node` is `null`,
-             starts from `state.path`. Only subattrsets that are JBoss
-             paths, i.e. follows the `key=value` format, are recursed
+          /* Recurses into the `nodeValue` attrset. Only subattrsets that
+             are JBoss paths, i.e. follows the `key=value` format, are recursed
              into - the rest are considered JBoss attributes / maps.
           */
-          recurse = state: node:
+          recurse = path: nodeValue:
             let
-              path = state.path ++ (lib.optional (node != null) node);
+              nodeValue' =
+                if builtins.isAttrs nodeValue && nodeValue._type or "" == "order" then nodeValue.content else nodeValue;
               isPath = name:
                 let
-                  value = lib.getAttrFromPath (path ++ [ name ]) attrs;
+                  value = nodeValue'.${name};
                 in
                   if (builtins.match ".*([=]).*" name) == [ "=" ] then
                     if builtins.isAttrs value || value == null then
                       true
                     else
-                      throw "Parsing path '${lib.concatStringsSep "." (path ++ [ name ])}' failed: JBoss attributes cannot contain '='!"
+                      throw "Parsing path '${concatStringsSep "." (path ++ [ name ])}' failed: JBoss attributes cannot contain '='!"
                   else
                     false;
-              jbossPath = "/" + (lib.concatStringsSep "/" path);
-              nodeValue = lib.getAttrFromPath path attrs;
-              children = if !builtins.isAttrs nodeValue then {} else nodeValue;
+              jbossPath = "/" + (concatStringsSep "/" path);
+              children = if !builtins.isAttrs nodeValue' then {} else nodeValue';
               subPaths = builtins.filter isPath (builtins.attrNames children);
-              jbossAttrs = lib.filterAttrs (name: _: !(isPath name)) children;
-            in
-              state // {
-                text = state.text + (
-                  if nodeValue != null then ''
+              subPaths' =
+                if all (name: children.${name}._type or "" != "order") subPaths
+                then subPaths
+                else
+                  let
+                    getPriority = name:
+                      let value = children.${name};
+                      in if value._type or "" == "order" then value.priority else 1000;
+                    compare = a: b: getPriority a < getPriority b;
+                  in sort compare subPaths;
```

```suggestion
              getPriority = name:
                let
                  value = children.${name};
                in
                  if value._type or "" == "order" then value.priority else 1000;
              orderedSubPaths = sort (a: b: getPriority a < getPriority b) subPaths;
```
Since we're practically always going to have ordered paths in the config, let's simplify this. Also, give it a meaningful name and keep the style consistent with the rest of the module.

---

Actually the only place we use it currently is when HTTPS is enabled, which I don't consider that widespread -- I expect most people to use reverse proxies instead. Also, even when it's enabled, the only two sub-attrsets that we use it is in root and `subsystem=elytron` subtrees.

Thinking that I copied this optimization from [modules library](https://github.com/NixOS/nixpkgs/blob/81cdd363b8b23dca69683141489c9eca7027e1c4/lib/modules.nix). Not sure how much it's still useful though.

I'm not sure what did you mean by keeping the style consistent, but I fixed several problems. Please review and point me at any leftover issues.

---

I wouldn't expect most people to bother setting up a reverse proxy when it works just fine without it, but maybe that's an incorrect assumption :shrug: I still don't think the optimization is worth the complexity, though.

By keeping the style consistent, I meant that both the first `if` in `subPaths'` and the `let` in `getPriority` had a different indentation and linebreak style to them than what's used in the rest of the module. I also move out `getPriority` to the outer `let` and inline `compare` to reduce the level of nesting.

On a related note, I was also thinking about making the module use nginx, since that would simplify things a bit and add support for ACME. Since all of this will change when Keycloak switches away from JBoss, that's probably a good point at which to restructure this.

---

Let's indeed simplify the code. I have also incorporated your style suggestions.